src/libraries/Common/src/System/Net/Http/aspnetcore/Quic/Implementations/Managed/Internal/QuicSocketContext.cs

```diff
@@ -0,0 +1,218 @@
+            // TODO-RZ: check if already running
+            _ = BackgroundWorker(_socketTaskCts.Token);
+        }
+
+        /// <summary>
+        ///     Used to signal the thread that one of the connections has data to send.
+        /// </summary>
+        internal void Ping()
+        {
+            _signalTcs.TrySetResult(0);
+        }
+
+        private ManagedQuicConnection? DispatchPacket(QuicReader reader, IPEndPoint remoteEp)
+        {
+            if (_client != null)
+            {
+                _client.ReceiveData(reader, remoteEp, DateTime.Now);
+                return _client;
+            }
+
+            ManagedQuicConnection? connection;
+            // we need to dispatch packets to appropriate connection based on the connection Id, so we need to parse the headers
+            byte first = reader.Peek();
+            if (HeaderHelpers.IsLongHeader(first))
+            {
+                if (!LongPacketHeader.Read(reader, out var header))
+                {
+                    // drop packet
+                    return null;
+                }
+
+                var connectionId = _connectionIds!.FindConnectionId(header.DestinationConnectionId);
+                if (connectionId == null)
+                {
+                    connectionId = new ConnectionId(header.DestinationConnectionId.ToArray());
+                    _connectionIds!.Add(connectionId!);
+                }
+
+                if (!_connections!.TryGetValue(connectionId!, out connection))
+                {
+                    connection = _connections![connectionId!] =
+                        new ManagedQuicConnection(_listenerOptions!, this, remoteEp);
+                }
+            }
+            else
+            {
+                if (!ShortPacketHeader.Read(reader, _connectionIds!, out var header))
+                {
+                    // drop packet
+                    return null;
+                }
+
+                connection = _connections![header.DestinationConnectionId];
+            }
+
+            reader.Seek(0);
+            bool connected = connection!.Connected;
+            connection.ReceiveData(reader, remoteEp, DateTime.Now);
+            // TODO-RZ: handle failed connection attempts
+            if (!connected && connection!.Connected)
+            {
+                // new connection established
+                _newConnections!.TryWrite(connection!);
+            }
+
+            return connection;
+        }
+
+
+        private async Task BackgroundWorker(CancellationToken token)
+        {
+            async Task SendData(ManagedQuicConnection c, QuicWriter w, Socket s, byte[] buf)
+            {
+                w.Reset(buf);
+                c.SendData(w, out var addr, DateTime.Now);
+                int written = w.BytesWritten + w.Buffer.Offset;
+                if (written > 0)
+                {
+                    await s.SendToAsync(new ArraySegment<byte>(buf, 0, written), SocketFlags.None, addr);
```

Comment:
You're probably missing `.ConfigureAwait(false)` here and probably in other places as well. Almost all `awaits` in .NET Core use them, info https://devblogs.microsoft.com/dotnet/configureawait-faq/.

Reply:
Yes, I was deferring adding them to alter, but there was no reason not to put them there anyway.
Also thanks for the link, I needed to refresh my knowledge about this stuff.